google-cloud-examples/src/main/java/com/google/cloud/examples/bigquery/snippets/JobSnippets.java

```diff
@@ -36,15 +41,62 @@ public JobSnippets(Job job) {
    * Example of waiting for a job until it reports that it is done.
    */
   // [TARGET isDone()]
-  public boolean isDone() {
+  public void isDone() throws InterruptedException {
+    // [START isDone]
+    while (!job.isDone()) {
+      Thread.sleep(1000L);
+    }
+    // [END isDone]
+  }
+
+  /**
+   * Example usage of {@code waitFor()}.
+   */
+  // [TARGET waitFor(WaitForOption...)]
+  public boolean waitFor() throws InterruptedException {
     try {
-      // [START isDone]
-      while (!job.isDone()) {
-        Thread.sleep(1000L);
+      // [START waitFor]
+      Job completedJob = job.waitFor();
+      if (completedJob == null) {
+        // job no longer exists
+      } else if (completedJob.status().error() != null) {
+        // job failed, handle error
+      } else {
+        // job completed successfully
       }
-      // [END isDone]
-    } catch (InterruptedException e) {
+      // [END waitFor]
+    } catch (TimeoutException e) {
+      // Timeouts shouldn't happen without a timeout option.
       return false;
+    } catch (BigQueryException e) {
+      return true;
+    }
+    return true;
+  }
+
+  /**
+   * Example usage of {@code waitFor()} with checking period and timeout.
+   */
+  // [TARGET waitFor(WaitForOption...)]
+  public boolean waitForWithOptions() throws InterruptedException {
+    try {
+      // [START waitFor]
+      Job completedJob =
+          job.waitFor(
+              WaitForOption.checkEvery(1, TimeUnit.SECONDS),
+              WaitForOption.timeout(60, TimeUnit.SECONDS));
+      if (completedJob == null) {
+        // job no longer exists
+      } else if (completedJob.status().error() != null) {
+        // job failed, handle error
+      } else {
+        // job completed successfully
+      }
+      // [END waitFor]
+    } catch (TimeoutException e) {
+      return true;
+    } catch (BigQueryException e) {
```

Comment:
You can also let `BigQueryException` go. They represent a failed test (e.g. the service is not recognizing a request). We throw in the test as well so that `verify` fails.